easybuild/tools/job/pbs_python.py

```diff
@@ -45,69 +47,122 @@
 # list of known hold types
 KNOWN_HOLD_TYPES = []
 
-pbs_import_failed = None
 try:
     from PBSQuery import PBSQuery
     import pbs
     KNOWN_HOLD_TYPES = [pbs.USER_HOLD, pbs.OTHER_HOLD, pbs.SYSTEM_HOLD]
+    # `pbs_python` available, no need guard against import errors
+    def only_if_pbs_import_successful(fn):
+        return fn
+    HAVE_PBS_PYTHON = True
 except ImportError:
-    _log.debug("Failed to import pbs from pbs_python. Silently ignoring, is only a real issue with --job")
-    pbs_import_failed = ("PBSQuery or pbs modules not available. "
-                         "Please make sure pbs_python is installed and usable.")
-
-
-def connect_to_server(pbs_server=None):
-    """Connect to PBS server and return connection."""
-    if pbs_import_failed:
-        _log.error(pbs_import_failed)
-        return None
-
-    if not pbs_server:
-        pbs_server = pbs.pbs_default()
-    return pbs.pbs_connect(pbs_server)
-
-
-def disconnect_from_server(conn):
-    """Disconnect a given connection."""
-    if pbs_import_failed:
-        _log.error(pbs_import_failed)
-        return None
-
-    pbs.pbs_disconnect(conn)
-
-
-def get_ppn():
-    """Guess the ppn for full node"""
-
-    log = fancylogger.getLogger('pbs_job.get_ppn')
-
-    pq = PBSQuery()
-    node_vals = pq.getnodes().values()  # only the values, not the names
-    interesting_nodes = ('free', 'job-exclusive',)
-    res = {}
-    for np in [int(x['np'][0]) for x in node_vals if x['state'][0] in interesting_nodes]:
-        res.setdefault(np, 0)
-        res[np] += 1
-
-    # return most frequent
-    freq_count, freq_np = max([(j, i) for i, j in res.items()])
-    log.debug("Found most frequent np %s (%s times) in interesting nodes %s" % (freq_np, freq_count, interesting_nodes))
-
-    return freq_np
+    _log.debug("Failed to import pbs from pbs_python."
+               " Silently ignoring, this is a real issue only with --job=pbs")
+    # no `pbs_python` available, turn function into a no-op
+    def only_if_pbs_import_successful(fn):
+        def instead(*args, **kwargs):
+            """This is a no-op since `pbs_python` is not available."""
+            errmsg = ("PBSQuery or pbs modules not available."
+                      " Please make sure `pbs_python` is installed and usable.")
+            _log.error(errmsg)
+            raise RuntimeError(errmsg)
+        return instead
+    HAVE_PBS_PYTHON = False
+
+
+class Pbs(JobServer):
+    """
+    Manage PBS server communication and create `PbsJob` objects.
+    """
+
+    USABLE = HAVE_PBS_PYTHON
+
+    def __init__(self, pbs_server=None):
+        self.pbs_server = pbs_server or pbs.pbs_default()
+        self.conn = None
+        self._ppn = None
+
+    def begin(self):
+        self.connect_to_server()
+        self._submitted = []
+
+    @only_if_pbs_import_successful
```

Comment:
in what case is the `USABLE` not sufficient to avoid mistakes? (USABEL was used to avoid decorators like this).